packages/airnode-node/src/coordinator/calls/chain-limits.ts

```diff
@@ -0,0 +1,78 @@
+import flatMap from 'lodash/flatMap';
+import groupBy from 'lodash/groupBy';
+import map from 'lodash/map';
+import zip from 'lodash/zip';
+import * as logger from '../../logger';
+import { sortRequests } from '../../requests';
+import {
+  Config,
+  ProviderStates,
+  Request,
+  GroupedRequests,
+  RequestType,
+  ApiCall,
+  Withdrawal,
+  PendingLog,
+} from '../../types';
+
+function groupRequests(flatRequests: Request<any>[]): GroupedRequests {
+  const apiCalls = flatRequests
+    .filter((request) => request.__type === RequestType.ApiCall)
+    .map((request) => request as Request<ApiCall>);
+
+  const withdrawals = flatRequests
+    .filter((request) => request.__type === RequestType.Withdrawal)
+    .map((request) => request as Request<Withdrawal>);
+
+  return { apiCalls, withdrawals };
+}
+
+export function applyChainLimits(config: Config, providerStates: ProviderStates) {
+  const logs: PendingLog[] = [];
+
+  const providerRequests = providerStates.evm
+    .map((evmState, index) => {
+      const { apiCalls, withdrawals } = evmState.requests;
+      return { index, chainId: evmState.settings.chainId, requests: sortRequests([...apiCalls, ...withdrawals]) };
+    })
+    // Sort by number of request ascendingly
+    .sort((p1, p2) => p1.requests.length - p2.requests.length);
+  const groupedProviderRequests = groupBy(providerRequests, 'chainId');
+  const chainRequests = map(groupedProviderRequests, (group, chainId) => ({
+    chainId,
+    chainLimit: config.chains.find((chain) => chain.id === chainId)!.maxConcurrentTransactions,
```

Comment:
This tries to group the providers by `chainId`. This assertion is indeed valid since every provider is listed in some under some chain settings inside [config.json](https://docs.api3.org/airnode/v0.4/grp-providers/tutorial/quick-deploy-aws/config-json.html). 

These providers are initialized one by one [here](https://github.com/api3dao/airnode/blob/an395-per-chain-limits/packages/airnode-node/src/providers/actions.ts#L43) and I am basically reverting that mapping.